### PR TITLE
community: update Memgraph integration

### LINKED ISSUE #25606: Unable to insert GraphDocuments into Memgraph
### Checked other resources

- [X] I added a very descriptive title to this issue.
- [X] I searched the LangChain documentation with the integrated search.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I am sure that this is a bug in LangChain rather than my code.
- [X] The bug is not resolved by updating to the latest stable version of LangChain (or the specific integration package).

### Example Code

```python
graph = MemgraphGraph(url="bolt://localhost:7687", username="", password="")
documents = [Document(page_content="this is test")]
llm_transformer = LLMGraphTransformer(llm=llm) # please define your llm here
graph_documents = llm_transformer.convert_to_graph_documents(documents)
graph.add_graph_documents(
    graph_documents,
    baseEntityLabel=False,
    include_source=False
)
```

### Error Message and Stack Trace (if applicable)

```python
ClientError: {code: Memgraph.ClientError.MemgraphError.MemgraphError} {message: There is no procedure named 'apoc.merge.node'.}
```

### Description

I understand that the community libs covers support for Memgraph database by extending the Neo4j graph class. However, the functions provided by the class methods do not work with Memgraph. For example. `add_graph_documents` throws multiple errors depending on the parameters set. In the code, setting parameters `baseEntityLabel` and `include_source` to `False` results in the following error

```python
ClientError: {code: Memgraph.ClientError.MemgraphError.MemgraphError} {message: There is no procedure named 'apoc.merge.node'.}
```

I'm not sure there exists any such procedures for memgraph as per this post on stackoverflow
https://stackoverflow.com/questions/75244143/is-there-something-in-memgraph-that-has-the-same-function-as-apoc-in-neo4j

Now, setting just the `baseEntityLabel` to `True` will result in the following error
```python
ClientError: {code: Memgraph.ClientError.MemgraphError.MemgraphError} {message: line 1:19 no viable alternative at input 'CREATECONSTRAINTIF'}
```

setting `include_source` to `True` results in the following error
```python
ClientError: {code: Memgraph.ClientError.MemgraphError.MemgraphError} {message: line 1:53 mismatched input 'text' expecting {ADD, ACTIVE, AFTER, ALTER, AT, ANALYZE, ASYNC, AUTH, BAD, BATCH_INTERVAL, BATCH_LIMIT, BATCH_SIZE, BEFORE, BUILD, BOOTSTRAP_SERVERS, CALL, CHECK, CLEAR, CLUSTER, COMMIT, COMMITTED, CONFIG, CONFIGS, CONSUMER_GROUP, COORDINATOR, CREATE_DELETE, CREDENTIALS, CSV, CURRENT, DATA, DISABLE, DO, DELIMITER, DEMOTE, DATABASE, DATABASES, DENY, DIRECTORY, DROP, DUMP, DURABILITY, EDGE, EDGE_TYPES, ENABLE, ENUM, ENUMS, EVERY, EXECUTE, FAILOVER, FOR, FORCE, FOREACH, FREE, FREE_MEMORY, FROM, GLOBAL, GRANT, GRAPH, GRANTS, HEADER, IDENTIFIED, IF, IGNORE, IMPORT, INACTIVE, IN_MEMORY_ANALYTICAL, IN_MEMORY_TRANSACTIONAL, INSTANCE, INSTANCES, ISOLATION, KAFKA, LABELS, LEVEL, LOAD, LOCK, MAIN, MODE, MODULE_READ, MODULE_WRITE, MULTI_DATABASE_EDIT, MULTI_DATABASE_USE, NEXT, NO, NODE_LABELS, NOTHING, OF_TOKEN, ON_DISK_TRANSACTIONAL, NULLIF, PASSWORD, PERIODIC, PORT, PRIVILEGES, PULSAR, READ, READ_FILE, REGISTER, REPLACE, REPLICA, REPLICAS, REPLICATION, RESET, REVOKE, ROLE, ROLES, ROWS, QUOTE, SCHEMA, SERVER, SERVICE_URL, SESSION, SETTING, SETTINGS, SNAPSHOT, START, STATE, STATISTICS, STATS, STATUS, STOP, STORAGE, STORAGE_MODE, STREAM, STREAMS, SYNC, TERMINATE, TIMEOUT, TO, TOPICS, TRANSACTION, TRANSACTION_MANAGEMENT, TRANSACTIONS, TRANSFORM, TRIGGER, TRIGGERS, TTL, UNCOMMITTED, UNLOCK, UNREGISTER, UPDATE, USE, USER, USERS, USING, VALUE, VALUES, VERSION, WEBSOCKET, ALL, AND, ANY, AS, ASC, ASCENDING, ASSERT, BFS, BY, CASE, COALESCE, CONSTRAINT, CONTAINS, COUNT, CREATE, CYPHERNULL, DELETE, DESC, DESCENDING, DETACH, DISTINCT, ELSE, END, ENDS, EXISTS, EXPLAIN, EXTRACT, FALSE, FILTER, IN, INDEX, INFO, IS, KB, KEY, LIMIT, L_SKIP, MATCH, MB, MEMORY, MERGE, NODE, NONE, NOT, ON, OPTIONAL, OR, ORDER, PROCEDURE, PROFILE, QUERY, REDUCE, REMOVE, RETURN, SET, SHOW, SINGLE, STARTS, THEN, TRUE, UNION, UNIQUE, UNLIMITED, UNWIND, WHEN, WHERE, WITH, WSHORTEST, ALLSHORTEST, XOR, YIELD, UnescapedSymbolicName, EscapedSymbolicName}}
```

Is Memgraph actually supported for usage with `LLMGraphTransformer`? Since `Neo4jGraph` can be used, and `MemgraphGraph` is derived from neo4j, i would expect it to work without any issue. Please let me know how i can use Memgraph with `LLMGraphTransformers`, if not can we add this support?

### System Info

System Information
------------------
> OS:  Darwin
> OS Version:  Darwin Kernel Version 22.3.0: Mon Jan 30 20:38:37 PST 2023; root:xnu-8792.81.3~2/RELEASE_ARM64_T6000
> Python Version:  3.11.9 (main, Apr  2 2024, 08:25:04) [Clang 15.0.0 (clang-1500.1.0.2.5)]

Package Information
-------------------
> langchain_core: 0.2.33
> langchain: 0.2.14
> langchain_community: 0.2.12
> langsmith: 0.1.84
> langchain_experimental: 0.0.64
> langchain_milvus: 0.1.4
> langchain_nvidia_ai_endpoints: 0.2.1
> langchain_openai: 0.1.22
> langchain_text_splitters: 0.2.2

Optional packages not installed
-------------------------------
> langgraph
> langserve

Other Dependencies
------------------
> aiohttp: 3.9.3
> async-timeout: Installed. No version info available.
> dataclasses-json: 0.6.7
> jsonpatch: 1.33
> numpy: 1.25.0
> openai: 1.41.1
> orjson: 3.10.6
> packaging: 24.1
> pillow: 10.4.0
> pydantic: 2.8.2
> pymilvus: 2.4.4
> PyYAML: 6.0.1
> requests: 2.31.0
> scipy: 1.14.0
> SQLAlchemy: 1.4.3
> tenacity: 8.5.0
> tiktoken: 0.7.0
> typing-extensions: 4.12.2

### PR DESCRIPTION
**Pull request by katarinasupe**
**Description:**
- **Memgraph** no longer relies on `Neo4jGraphStore` but **implements `GraphStore`**, just like other graph databases.
- **Memgraph** no longer relies on `GraphQAChain`, but implements `MemgraphQAChain`, just like other graph databases.
- The refresh schema procedure has been updated to try using `SHOW SCHEMA INFO`. The fallback uses Cypher queries (a combination of schema and Cypher) → **LangChain integration no longer relies on MAGE library**.
- The **schema structure** has been reformatted. Regardless of the procedures used to get schema, schema structure is the same.        
- The `add_graph_documents()` method has been implemented. It transforms `GraphDocument` into Cypher queries and creates a graph in Memgraph. It implements the ability to use `baseEntityLabel` to improve speed (`baseEntityLabel` has an index on the `id` property). It also implements the ability to include sources by creating a `MENTIONS` relationship to the source document.
- Jupyter Notebook for Memgraph has been updated.
- **Issue:** /
- **Dependencies:** /
- **Twitter handle:** supe_katarina (DX Engineer @ Memgraph)

Closes #25606

### CODE REVIEW

**Comment by katarinasupe** [comment]
Hi @baskaryan, tagging as noted in PR template. Please let me know what I need to do regarding the test and what's needed to do in order to get this merged.

**Comment by katarinasupe** [comment]
I managed to fix the CI. Now, all is passing.

Tagging @efriis, as the other person from the PR template for the review. Let me know if someone else should be tagged.

**Comment by efriis** [comment]
Hey! Here's a link to the review policy: https://python.langchain.com/docs/contributing/review_process/

Status is currently needs support

**Comment by katarinasupe** [comment]
@efriis is there any update? I have more than 5 upvotes so I am wondering what's the next course of action here and if I can help somehow.

**Review by libraua — APPROVED** [review]

**Comment by katarinasupe** [comment]
Thank you, @libraua, for approving this PR. Is there something else I need to do to get this PR merged? I wanted to enable Memgraph developers to utilize Memgraph LangChain integration with the latest improvements.

File: libs/community/langchain_community/chains/graph_qa/memgraph.py [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+                )
+                qa_chain = response_prompt | qa_llm | StrOutputParser()  # type: ignore
+            except (NotImplementedError, AttributeError):
+                raise ValueError("Provided LLM does not support native tools/functions")
+        else:
+            qa_chain = LLMChain(llm=qa_llm, **use_qa_llm_kwargs)  # type: ignore[arg-type]
+
+        cypher_generation_chain = LLMChain(
+            llm=cypher_llm or llm,  # type: ignore[arg-type]
+            **use_cypher_llm_kwargs,  # type: ignore[arg-type]
```
Comment (ccurme) [inline libs/community/langchain_community/chains/graph_qa/memgraph.py:237]: do we need `use_cypher_llm_kwargs` or `use_qa_llm_kwargs`? Ideally we would avoid LLMChain and just use `prompt | llm | parser` (e.g., as you do above).

**Review by ccurme — COMMENTED** [review]
Hi @katarinasupe, I don't want to block this but have some qualms about `MemgraphQAChain`. I think the fact that other graph databases implement custom `Chain` classes is not an advantage and may not be desirable to emulate. `MemgraphQAChain` also uses LLMChain, which is deprecated (I include a comment below).

LangChain is investing in LangGraph, and instead of releasing and maintaining a v0.0 abstraction, you could advantage yourself by implementing these as [tools](https://python.langchain.com/docs/concepts/tools/) and adding documentation demonstrating their use in question-answering using a modern agent framework.

- Migration guides for other chains: https://python.langchain.com/docs/versions/migrating_chains/
- Example of how we'd recommend querying SQL data using an agent: https://python.langchain.com/docs/integrations/tools/sql_database/#use-within-an-agent

Our documentation across integrations has not kept up with our current understanding of best practices, so apologies for the confusion. If you want to move forward with `MemgraphQAChain` we can go ahead (but suggest we remove LLMChain). Wanted to make sure you're aware of current best practices first.

File: libs/community/langchain_community/chains/graph_qa/memgraph.py (hunk 1 above)
Comment (katarinasupe) [reply libs/community/langchain_community/chains/graph_qa/memgraph.py:237]: Thank you for the review. Let me digest this today and figure out the next best steps 🙏

Comment (katarinasupe) [reply libs/community/langchain_community/chains/graph_qa/memgraph.py]: Hi @ccurme any updates here? I would love to get this merged if possible 🙏

**Review by ccurme — APPROVED** [review]

**Comment by pthalasta** [comment]
@ccurme is this available as part of langchain-core==0.3.23?

**Comment by ccurme** [comment]
This will be available in the next release of `langchain-community` (0.3.11)

**Comment by pthalasta** [comment]
@katarinasupe while trying out these changes i do encounter the following error when i try to insert the graph documents into memgraph. I'm not sure if this PR handled this and if i'm on a different version of lib than that you tested on.

```
ClientError: {code: Memgraph.ClientError.MemgraphError.MemgraphError} {message: There is no procedure named 'merge.node'.}```

**Comment by katarinasupe** [comment]
Hi @pthalasta, that can happen if you're using only Memgraph without the MAGE library. Please install [Memgraph MAGE](https://memgraph.com/docs/advanced-algorithms/install-mage) and let me know if that works. If you're still experiencing issues, you can [join our Discord](https://discord.gg/memgraph) or book [an office hours](https://memgraph.com/office-hours) to ask questions.

**Pull request merged**